chalice/cli/__init__.py

```diff
@@ -167,32 +149,67 @@ def local(ctx, port=8000):
               default=True,
               help='Automatically generate IAM policy for app code.')
 @click.option('--profile', help='Override profile at deploy time.')
-@click.argument('stage', nargs=1, required=False)
+@click.option('--api-gateway-stage',
+              help='Name of the API gateway stage to deploy to.')
+@click.option('--stage', default='dev',
+              help=('Name of the Chalice stage to deploy to. '
+                    'Specifying a new chalice stage will create '
+                    'an entirely new set of AWS resources.'))
+@click.argument('deprecated-api-gateway-stage', nargs=1, required=False)
 @click.pass_context
-def deploy(ctx, autogen_policy, profile, stage):
-    # type: (click.Context, bool, str, str) -> None
-    config = create_config_obj(
-        # Note: stage_name is not the same thing as the chalice stage.
-        # This is a legacy artifact that just means "API gateway stage",
-        # or for our purposes, the URL prefix.
-        ctx, stage_name='dev', autogen_policy=autogen_policy,
-        profile=profile)
-    if stage is None:
-        stage = 'dev'
-    session = create_botocore_session(profile=config.profile,
-                                      debug=ctx.obj['debug'])
-    d = deployer.create_default_deployer(session=session, prompter=click)
-    try:
-        deployed_values = d.deploy(config, stage_name=stage)
-        record_deployed_values(deployed_values, os.path.join(
-            config.project_dir, '.chalice', 'deployed.json'))
-    except botocore.exceptions.NoRegionError:
-        e = click.ClickException("No region configured. "
-                                 "Either export the AWS_DEFAULT_REGION "
-                                 "environment variable or set the "
-                                 "region value in our ~/.aws/config file.")
-        e.exit_code = 2
-        raise e
+def deploy(ctx, autogen_policy, profile, api_gateway_stage, stage,
+           deprecated_api_gateway_stage):
+    # type: (click.Context, bool, str, str, str, str) -> None
+    if api_gateway_stage is not None and \
+            deprecated_api_gateway_stage is not None:
+        raise _create_deprecated_stage_error(api_gateway_stage,
+                                             deprecated_api_gateway_stage)
+    if api_gateway_stage is None and deprecated_api_gateway_stage is not None:
```

Comment:
Updated, that was before adding the mutually exclusive part validation so the first clause isn't needed anymore.